### ISSUE TITLE
Revise, document and test patch to persp() for better x-axis labelling when x supplied as a list

### ISSUE THREAD

**Issue by hturner** [r-devel/r-dev-day#37]
This issue picks up on [Bug 1161](https://bugs.r-project.org/show_bug.cgi?id=1161) that we discussed at an R Contributor Office Hour. Following the discussion, @mcheema and I iterated on a patch and the task here is to revise the patch in response to review from @pmur002.

Suggested next  steps:

- [ ] Update the help file (persp.Rd) to say that, if 'z' is not specified, 'x' can be a matrix (used for 'z') OR 'x' can be a list (used for 'x', 'y', and 'z').
- [ ] Review changes to code (persp.R) 
    - The code for defaulting 'xlab', 'ylab', and 'zlab' will be difficult to maintain because it has to "mirror" the code that defaults 'x', 'y', and 'z'. Perhaps add at least a comment to indicate that any changes to one need to be matched by changes to the other.
    - The code in xy.coords() and xyz.coords() that does something similar will error if the list does not have 'x' and 'y' and 'z') components. persp() could do similar, though that might have negative impacts on existing code.
- [ ] This is a candidate for visual regression testing (e.g., with {gdiff}). Test examples(persp) and the output from demo(persp).
- [ ] Could also search for uses of persp() on CRAN (via [CRAN "mirror" on GitHub](https://github.com/cran)) to check potential impact and possibly visual regression test those. If that were done and the results were clean, that would allow for more ambitious errors on malformed lists (see above).

For searching on GitHub, note that you can [search with regular expressions](https://docs.github.com/en/search-github/github-code-search/understanding-github-code-search-syntax#using-regular-expressions) to narrow the results.

This will give an introduction to patching code and documentation and testing the impact of a code patch, without needing to get deep into the code (it only touches R code). Good for a new contributor, perhaps with some R package development experience.

I am happy to work with someone on this, or for others to take over from here. Maybe we can form a small Graphics team, with people working on this and #36, which is a similar type of issue.

**Comment by milanmlft** [comment]
> The code in xy.coords() and xyz.coords() that does something similar will error if the list does not have 'x' and 'y' and 'z') components. persp() could do similar, though that might have negative impacts on existing code.

I think the current version of `persp()` also throws an error if the list doesn't have `x`, `y` and `z`, so I think that should be fine?

```r
persp(list(a = x1, b = y1, c = z1))

#> Error in `persp.default()`:
#> ! invalid 'x' argument
```

**Comment by milanmlft** [comment]
My progress on updating the documentation 

[persp-Rd.patch](https://github.com/user-attachments/files/16894933/persp-Rd.patch)

**Comment by pmur002** [comment]
Good point that it already errors, though the error message could be clearer about what is wrong with `x`

**Comment by njtierney** [comment]
@pmur002 should it be something like "x must be a list with unique names starting with 'x', 'y', and 'z' "

The error starts here: https://github.com/r-devel/r-svn/blob/cc70121992667e41c57ac5a496ed553fad4b92d4/src/library/graphics/R/persp.R#L39-L43

```r
#...
            if (is.list(x)) {
                z <- x$z
                y <- x$y
                x <- x$x
            }
#...
```

Since this will return `NULL` which then causes it to fail here:

https://github.com/r-devel/r-svn/blob/cc70121992667e41c57ac5a496ed553fad4b92d4/src/library/graphics/R/persp.R#L66-L70

```r
    r <- .External.graphics(C_persp, x, y, z, xlim, ylim, zlim, theta, phi, r, d,
                            scale, expand, col, border, ltheta, lphi, shade,
                            box, axes, nticks, ticktype,
                            as.character(xlab), as.character(ylab),
                            as.character(zlab), ...)
```

We could perhaps add some error checking in here, something of the flavour of:

```r
#...
##
x_names <- names(x)
right_x_names <- all(x_names %in% c("x", "y", "z"))
if (!right_x_names) {
  stop("`x` must have names 'x', 'y', 'z', we saw:", x_names, .call = FALSE)
}
##
            if (is.list(x)) {
                z <- x$z
                y <- x$y
                x <- x$x
            }
#...
```

**Comment by ashishtiwari03** [comment]
![](vscode-remote://codespaces%2Bsolid-space-orbit-r49pp79q4r2x74/workspaces/r-dev-env/test_list.png)
Tested Milan's persp.Rd patch on current trunk (dev build via $BUILDDIR/bin/R):

Patch: applies cleanly (U), no conflicts. Rebuilds fine; ?persp now renders the list documentation on the z argument correctly.
List input works: persp(list(x = x, y = y, z = z)) draws a correct surface (image attached).
x-axis labelling (the focus of this issue): when x is supplied as a list, the default x-label is the full expression list(x = x, y = y, z = z) rather than a clean label, so the labelling improvement doesn't appear to be in trunk yet.
Malformed list error: persp(list(a = x, b = y, c = z)) gives Error ... : invalid 'x' argument plus several no non-missing arguments to min/max warnings, i.e. the clearer error message discussed above (requiring names x, y, z) isn't in yet.

So docs + basic list input are good; the x-label fix and the improved error message still look outstanding.

**Comment by ashishtiwari03** [comment]
Implemented both fixes in persp.R (patch attached, against r90173):

Labelling: list input now labels axes name$x/$y/$z, x-axis reads xyz$x (image attached), matching the bug 1161 proposal, instead of the full list(...) expression.
Malformed list: following njtierney's check with Mächler's condition all(c("x","y","z") %in% names(x)), a bad list now errors with 'x' must be a list with components 'x', 'y', and 'z' instead of invalid 'x' argument + min/max warnings.
Regression: normal non-list calls (persp(x, y, z)) unchanged.

Patch attached inline below (against r90173):

```diff
Index: src/library/graphics/R/persp.R
===================================================================
--- src/library/graphics/R/persp.R	(revision 90173)
+++ src/library/graphics/R/persp.R	(working copy)
@@ -27,6 +27,12 @@
             col = "white", border = NULL, ltheta = -135, lphi = 0, shade = NA,
             box = TRUE, axes = TRUE, nticks = 5, ticktype = "simple", ...)
 {
+    if (!missing(x) && is.list(x)) {
+        xn <- deparse1(substitute(x))
+        if (is.null(xlab)) xlab <- paste0(xn, "$x")
+        if (is.null(ylab)) ylab <- paste0(xn, "$y")
+        if (missing(z) && is.null(zlab)) zlab <- paste0(xn, "$z")
+    }
     if (is.null(xlab))
         xlab <- if (!missing(x)) deparse1(substitute(x)) else "X"
     if (is.null(ylab))
@@ -37,6 +43,8 @@
     if (missing(z)) {
         if (!missing(x)) {
             if (is.list(x)) {
+                if(!all(c("x","y","z")%in% names(x)))
+                   stop("'x' must be a list with components 'x', 'y', and 'z'")
                 z <- x$z
                 y <- x$y
                 x <- x$x
```
Happy to adjust the label style if you'd prefer a different format.
